Make mjpeg handling work again

timbms opened the pull request:
Accept initial multipart response , accept subsequent JPEG frames
Processing previous JPEG frame
Decode MJPEG frames
Fix for #950 and #999

timbms committed:
Make mjpeg handling work again

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
Merge branch 'develop' into mjpegHandling

timbms committed:
Removed unnecessary logging
Change on aspectRatio only when necessary
Removed excessive logging

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

digitaldan commented on OpenHABCore/Sources/OpenHABCore/Util/HTTPClient.swift:
```diff
@@ -360,7 +386,82 @@ public final class HTTPClient: NSObject, Sendable {
+
+        // Add Basic Auth header if credentials are available
+        if !connectionConfiguration.username.isEmpty, !connectionConfiguration.password.isEmpty {
+            let credentials = "\(connectionConfiguration.username):\(connectionConfiguration.password)"
```
Shouldn't credential handling be done in a delegate, callback or something similar?  what about if the user is using Client certs for example?    Can this just use what we already have for that?

timbms replied:
Agree. See new commit

digitaldan commented on OpenHABCore/Sources/OpenHABCore/Util/MJPEGStreamDelegate.swift:
```diff
@@ -0,0 +1,487 @@
+        }
+    }
+
+    func urlSession(_ session: URLSession, task: URLSessionTask, didReceive challenge: URLAuthenticationChallenge) async -> (URLSession.AuthChallengeDisposition, URLCredential?) {
```
I did not see this before making  my other comment, is it possible to share this with our other http clients auth handler so there's only one place in the code that handles auth?  Also what about client certs?

timbms replied:
Right I now delegate the authentication to HTTPClientDelegate

timbms committed:
Replace authentication method to delegate to HTTPClientDelegate
Integrate certificate handling

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
simplifying the mjpeg capturing

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
Cleaning up

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
Use HTTPClient.session.dataTask() instead of processStream()
- Keep the  which handles the streaming data properly
- Set the correct Accept header for MJPEG streams
- Use the HTTPClient's session which has the proper authentication and certificate handling

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms commented on OpenHABCore/Sources/OpenHABCore/Util/SimpleMJPEGStreamDelegate.swift:
```diff
@@ -0,0 +1,86 @@
+import os
+import UIKit
+
+public final class SimpleMJPEGStreamDelegate: NSObject, URLSessionDataDelegate, URLSessionTaskDelegate, @unchecked Sendable {
+    private var imageData = Data()
+    private var isFirstFrame = true
```
Thanks

timbms committed:
Clean-up

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms commented:
@digitaldan @TAKeanice much easier to review because it is become so much shorter in the next iteration

timbms committed:
Switching from ScaleAspectFitImageView to regular UIImageView - This eliminates the dual aspect ratio constraint management that was causing timing conflicts
Setting lower priority (998) for manual aspect ratio constraints - This ensures they don't conflict with any remaining system constraints
Always applying initial 16:9 ratio - Provides consistent initial sizing regardless of load vs reload
Improved constraint lifecycle management

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
Stream Continuity: MJPEG streams persist across sitemap reloads since they're managed outside the table view cells
VideoStreamManager singleton manages video streams independently of table view cell lifecycle
Modified SimpleMJPEGPlayer to support updating the target imageView without recreating the stream
Updated VideoUITableViewCell to use the shared VideoStreamManager for MJPEG streams instead of creating individual streams
Added proper cleanup with reference counting to ensure streams are only stopped when no cells are using them

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms commented on openHAB/VideoUITableViewCell.swift:
```diff
@@ -95,23 +95,70 @@ class VideoUITableViewCell: GenericUITableViewCell, NoIconDisplayableCell {
+
+            if let newUrl {
+                currentStreamUrl = newUrl
+                mjpegPlayer = VideoStreamManager.shared.getOrCreateStream(
```
Thanks

timbms committed:
Added tests for SimpleMJPEGStreamDelegate
Straighter initializers for HTTIPClient

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
Added proper view state management:
Hide playerView and show mainImageView for MJPEG videos
Hide mainImageView and show playerView for HLS videos
Clear opposite video type when switching
Enhanced prepareForReuse(): Proper cleanup of all video state when cell is reused
Added initial view state: Both views start hidden until needed

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>
